helix-rest/src/test/java/org/apache/helix/rest/server/TestAuthValidator.java

```diff
@@ -0,0 +1,143 @@
+    put("clusters/" + testClusterName, null, Entity.entity("", MediaType.APPLICATION_JSON_TYPE),
+        Response.Status.CREATED.getStatusCode());
+    String body = get("clusters/", null, Response.Status.OK.getStatusCode(), true);
+    JsonNode node = OBJECT_MAPPER.readTree(body);
+    String clustersStr = node.get(ClusterAccessor.ClusterProperties.clusters.name()).toString();
+    Assert.assertTrue(clustersStr.contains(testClusterName));
+  }
+
+  @Test(dependsOnMethods = "testDefaultAuthValidator")
+  public void testCustomAuthValidator() throws IOException, InterruptedException {
+    String testClusterName = "testCustomAuthValidator";
+    int newPort = getBaseUri().getPort() + 1;
+
+    // Start a second server for testing Distributed Leader Election for writes
+    _mockBaseUri = HttpConstants.HTTP_PROTOCOL_PREFIX + getBaseUri().getHost() + ":" + newPort;
+    _httpClient = HttpClients.createDefault();
+
+    AuthValidator mockAuthValidatorPass = Mockito.mock(AuthValidator.class);
+    when(mockAuthValidatorPass.validate(any())).thenReturn(true);
+    AuthValidator mockAuthValidatorReject = Mockito.mock(AuthValidator.class);
+    when(mockAuthValidatorReject.validate(any())).thenReturn(false);
+
+    List<HelixRestNamespace> namespaces = new ArrayList<>();
+    namespaces.add(new HelixRestNamespace(HelixRestNamespace.DEFAULT_NAMESPACE_NAME,
+        HelixRestNamespace.HelixMetadataStoreType.ZOOKEEPER, ZK_ADDR, true));
+
+    // Create a server that allows operations based on namespace auth and rejects operations based
+    // on cluster auth
+    HelixRestServer server =
+        new HelixRestServer(namespaces, newPort, getBaseUri().getPath(), Collections.emptyList(),
+            mockAuthValidatorReject, mockAuthValidatorPass);
+    server.start();
+
+    HttpUriRequest request =
+        buildRequest("/clusters/" + testClusterName, HttpConstants.RestVerbs.PUT, "");
+    sendRequestAndValidate(request, Response.Status.CREATED.getStatusCode());
+    request = buildRequest("/clusters/" + testClusterName, HttpConstants.RestVerbs.GET, "");
+    sendRequestAndValidate(request, Response.Status.FORBIDDEN.getStatusCode());
+
+    server.shutdown();
+    _httpClient.close();
+
+    // Create a server that rejects operations based on namespace auth and allows operations based
+    // on cluster auth
+    server =
+        new HelixRestServer(namespaces, newPort, getBaseUri().getPath(), Collections.emptyList(),
+            mockAuthValidatorPass, mockAuthValidatorReject);
+    server.start();
+    _httpClient = HttpClients.createDefault();
+
+    request = buildRequest("/clusters/" + testClusterName, HttpConstants.RestVerbs.GET, "");
+    sendRequestAndValidate(request, Response.Status.OK.getStatusCode());
+    request = buildRequest("/clusters", HttpConstants.RestVerbs.GET, "");
+    sendRequestAndValidate(request, Response.Status.FORBIDDEN.getStatusCode());
+
+    server.shutdown();
+    _httpClient.close();
```

Comment:
why do we need to keep closing the `_httpClient`?

Reply:
Apparently, if I don't close it here, the 2nd call below this line will hang indefinitely. No idea why and I didn't have time to dig deep. I wanted to ask you too: do you recall something like this happening before during MSDS testing?

Reply:
I don't remember running into this issue, but I think for testing, this shouldn't be a big deal.